File: novaact/worker/run_scope.py
"""Nova Act 薄 worker（ADR 0022/0024）：读 stdin 的 job JSON → 跑一个 scope → 吐 0024 事件到 stdout。

这是从 bdd/test_generic_steps.py 改造而来：脱掉 pytest-bdd 装饰器，逻辑（开会话/act/投票/派发）原样复用。
core 经子进程 adapter 起本 worker（ADR 0026 机制层），讲 0024 协议。

一生（ADR 0024）：
  读 stdin job → 开 AgentCore 会话 → 按 scope 串行跑 scenarios（每 step 派发）→ 逐事件吐 stdout
  → scope_done → 退出。捕获 SIGTERM：finally 停会话（终止契约）。

派发（ADR 0020/0024）：
  step.text 含 URL 字面量（引号内 https?://）→ 内建确定性导航 go_to_url（不浪费 AI）
  keyword=When → act（AI 动作，无 votes）
  keyword=Then → act_get(BOOL) + N 次投票（AI 断言，带 votes）
  keyword=Given 且非 URL → 也走 act（前置动作）

cost（ADR 0024）：Nova Act 按 agent-hour 计费，cost_usd = time_worked_s/3600 × rate（默认 4.75）。

跑（一般由 core adapter spawn，也可手动）：
  echo '<job json>' | AWS_REGION=us-east-1 .venv/bin/python worker/run_scope.py
"""
from __future__ import annotations

import json
import os
import re
import signal
import sys
from pathlib import Path

from nova_act import NovaAct, AgentCoreBrowserSessionProvider, BOOL_SCHEMA, Workflow
from nova_act.types.workflow import set_current_workflow, get_current_workflow

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))  # novaact/ 根，便于 import lib
from lib.workflow_setup import ensure_workflow_definition

REGION = os.environ.get("AWS_REGION", "us-east-1")
MODEL_ID = "nova-act-latest"
WORKFLOW_DEF = "spike-wikipedia-benchmark"
VOTES = 3  # AI 断言投票次数（治种类A抖动，ADR 0014）
NOVA_USD_PER_AGENT_HOUR = 4.75  # ADR 0024，来源 aws.amazon.com/nova/pricing

_URL_IN_QUOTES = re.compile(r'"(https?://[^"]+)"')

# 三通道分离（ADR 0024）：0024 事件走专用 fd（core adapter 读这个），与 SDK 打到 stdout 的进度噪声、
# worker 自己的诊断（stderr）物理隔离。adapter 经环境变量 EVENTS_FD 告知该 fd 号（pass_fds 继承，号不固定）。
# 无 EVENTS_FD（手动直跑、无 adapter）时回落 stdout，便于调试（`echo job | worker` 仍能看事件）。
_events_fd = os.environ.get("EVENTS_FD")
try:
    _events_out = os.fdopen(int(_events_fd), "w", encoding="utf-8") if _events_fd else sys.stdout
except (OSError, ValueError):
    _events_out = sys.stdout


def emit(obj: dict) -> None:
    """吐一条 0024 事件到事件通道（fd3，JSON Lines，字段名 camelCase 与 core/wire.py 一致）。"""
    _events_out.write(json.dumps(obj, ensure_ascii=False) + "\n")
    _events_out.flush()


def log(msg: str) -> None:
    sys.stderr.write(f"{msg}\n")
    sys.stderr.flush()


def _cost_from_result(r) -> dict | None:
    """从 act/act_get 结果的 metadata.time_worked_s 算 agent_time cost（ADR 0024）。"""
    md = getattr(r, "metadata", None)
    if md is None:
        return None
    tw = getattr(md, "time_worked_s", None)
    if tw is None:
        return None
    cost_usd = tw / 3600.0 * NOVA_USD_PER_AGENT_HOUR
    return {
        "cost_usd": round(cost_usd, 6),
        "precision": "estimated",  # agent_time → estimated（ADR 0024）
        "basis": "agent_time",
        "evidence": {
            "time_worked_s": tw,
            "human_wait_time_s": getattr(md, "human_wait_time_s", 0.0),
            "num_steps_executed": getattr(md, "num_steps_executed", None),
        },
    }


def _run_step(nova, scenario_id: str, step: dict) -> str:
    """派发执行一个 step，吐 step_done 事件，返回该 step 的 status（passed/failed/error）。"""
    idx = step["index"]
    keyword = step["keyword"]
    text = step["text"]

    try:
        url_match = _URL_IN_QUOTES.search(text)
        if url_match:
            # 内建确定性导航（ADR 0020）：抽 URL 直接 go_to_url，不浪费 AI
            nova.go_to_url(url_match.group(1))
            emit({"type": "step_done", "scenarioId": scenario_id, "stepIndex": idx, "status": "passed"})
            return "passed"

        if keyword == "Then":
            # AI 断言 + N 次投票（ADR 0014/0024）
            votes = []
            last_cost = None
            for _ in range(VOTES):
                r = nova.act_get(_unquote(text), BOOL_SCHEMA)
                votes.append(bool(r.matches_schema and r.parsed_response))
                last_cost = _cost_from_result(r)
            yes = sum(votes)
            passed = yes > VOTES / 2
            ev = {
                "type": "step_done", "scenarioId": scenario_id, "stepIndex": idx,
                "status": "passed" if passed else "failed",
                "votes": {"yes": yes, "total": VOTES},
            }
            if last_cost:
                ev["cost"] = last_cost
            if not passed:
                ev["errorType"] = "assertion_failed"
                ev["message"] = f"AI 断言未过多数票（{yes}/{VOTES}）：{text}"
            emit(ev)
            return "passed" if passed else "failed"

        # When / Given（非 URL）→ AI 动作（无 votes）
        r = nova.act(_unquote(text))
        ev = {"type": "step_done", "scenarioId": scenario_id, "stepIndex": idx, "status": "passed"}
        cost = _cost_from_result(r)
        if cost:
            ev["cost"] = cost
        emit(ev)
        return "passed"

    except Exception as e:
        emit({
            "type": "step_done", "scenarioId": scenario_id, "stepIndex": idx,
            "status": "error", "errorType": "engine_error", "message": f"{type(e).__name__}: {e}",
        })
        return "error"


def _unquote(text: str) -> str:
    """step 人话外层若整体被引号包裹（QA 写 When "搜索 X"），剥掉外引号喂引擎。"""
    t = text.strip()
    if len(t) >= 2 and t[0] == '"' and t[-1] == '"':
        return t[1:-1]
    return t


def _aggregate(statuses: list[str]) -> str:
    if any(s == "error" for s in statuses):
        return "error"
    if any(s == "failed" for s in statuses):
        return "failed"
    return "passed"


def main() -> int:
    job = json.loads(sys.stdin.readline())
    scope = job["scope"]
    scenarios = job["scenarios"]

    nova = None
    session_id = None

    def _on_sigterm(signum, frame):
        # 终止契约（ADR 0024）：收到 SIGTERM → 关会话再退。with 块的 __exit__ 在 sys.exit 时不一定跑，
        # 故这里显式关。
        log("worker: SIGTERM received, closing AgentCore session")
        try:
            if nova is not None:
                nova.close()
        except Exception:
            pass
        sys.exit(0)

    signal.signal(signal.SIGTERM, _on_sigterm)

    ensure_workflow_definition(WORKFLOW_DEF, region=REGION, description="Nova Act worker (ADR 0024)")
    wf = Workflow(model_id=MODEL_ID, boto_session_kwargs={"region_name": REGION}, workflow_definition_name=WORKFLOW_DEF)
    with wf:
        outer = get_current_workflow()
        set_current_workflow(wf)
        try:
            provider = AgentCoreBrowserSessionProvider(region=REGION)
            with provider.cdp_session() as (ws_url, headers):
                with NovaAct(
                    cdp_endpoint_url=ws_url, cdp_headers=headers, browser_auth=provider,
                    starting_page="about:blank",
                ) as nova:
                    # 取真实 AgentCore 会话 id（血缘，进 scope_done → RunStore，ADR 0016/0024）。
                    # NovaAct 已 started，get_session_id() 安全。
                    session_id = nova.get_session_id()
                    # scope 内串行跑 scenarios，共享同一会话（ADR 0019/0024）
                    for sc in scenarios:
                        sid = sc["id"]
                        emit({"type": "scenario_started", "scenarioId": sid})
                        statuses = [_run_step(nova, sid, st) for st in sc["steps"]]
                        emit({"type": "scenario_done", "scenarioId": sid, "status": _aggregate(statuses)})
        finally:
            set_current_workflow(outer)

    emit({
        "type": "scope_done", "scopeId": scope["id"], "sessionId": session_id,
        "costRate": {"nova_act_usd_per_agent_hour": NOVA_USD_PER_AGENT_HOUR},
    })
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: midscene/worker/run-scope.ts
// Midscene 薄 worker（ADR 0022/0024）：读 stdin 的 job JSON → 跑一个 scope → 吐 0024 事件到事件通道。
//
// 从 bdd/steps/generic.steps.ts 改造而来：脱掉 cucumber 装饰器，逻辑（开会话/aiAct/aiBoolean 投票/派发）原样复用。
// core 经子进程 adapter 起本 worker（ADR 0026 机制层），讲 0024 协议——与 Nova Act 腿对称。
//
// 三通道分离（ADR 0024）：0024 事件吐到 EVENTS_FD 指定的 fd（无则回落 stdout，便于手动直跑调试）；
// Midscene/SDK 的 stdout 噪声留 stdout；worker 自身诊断走 stderr。
//
// cost（ADR 0024）：从 agent._unstableLogContent() 的 executions[].tasks[].usage.total_tokens 取原生
// token 数，worker 只报 {tokens}；core 合计、美元折算交消费者（不追 Qwen 单价）。两腿对称：都只报原生量。
//
// 跑（一般由 core adapter spawn；也可手动）：
//   echo '<job json>' | AWS_REGION=us-east-1 node --import tsx/esm worker/run-scope.ts
import OpenAI from "openai";
import { PlaywrightAgent } from "@midscene/web/playwright";
import { chromium, type Browser } from "playwright";
import {
  BedrockAgentCoreClient,
  StartBrowserSessionCommand,
  StopBrowserSessionCommand,
} from "@aws-sdk/client-bedrock-agentcore";
import * as fs from "node:fs";
import { sigv4Fetch, signCdpUpgrade, BASE_URL, MODEL, REGION } from "../lib/agentcore-sigv4.mjs";

const BROWSER_ID = "aws.browser.v1";
const VOTES = 3; // AI 断言投票次数（治种类A抖动，ADR 0014）
const MODEL_CONFIG = {
  MIDSCENE_MODEL_NAME: MODEL,
  MIDSCENE_MODEL_BASE_URL: BASE_URL,
  MIDSCENE_MODEL_API_KEY: "unused",
  MIDSCENE_USE_QWEN3_VL: "true",
};
const URL_IN_QUOTES = /"(https?:\/\/[^"]+)"/;

// --- 事件通道：写 EVENTS_FD（adapter 经环境变量告知 fd 号，pass_fds 继承）；无则回落 stdout ---
const EVENTS_FD = process.env.EVENTS_FD ? Number(process.env.EVENTS_FD) : 1;
function emit(obj: unknown): void {
  fs.writeSync(EVENTS_FD, JSON.stringify(obj) + "\n"); // 同步写：保证事件按序到达（ADR 0024 顺序不变量）
}
function log(msg: string): void {
  process.stderr.write(msg + "\n");
}

interface Step { index: number; keyword: string; text: string; argument?: unknown }
interface Scenario { id: string; name: string; steps: Step[] }
interface Job { scope: { id: string; name: string }; engine: string; scenarios: Scenario[] }

function unquote(text: string): string {
  const t = text.trim();
  return t.length >= 2 && t.startsWith('"') && t.endsWith('"') ? t.slice(1, -1) : t;
}

// 报 Midscene/Bedrock 原生量 token（ADR 0024：engine 只报原生量，core 不算美元）。
// 从 agent._unstableLogContent 取最近一次 AI 调用的 usage.total_tokens；取不到则返回 undefined。
function lastCost(agent: PlaywrightAgent): Record<string, unknown> | undefined {
  try {
    const content = (agent as any)._unstableLogContent?.();
    const execs = content?.executions ?? [];
    let usage: any = undefined;
    for (const ex of execs) for (const task of ex.tasks ?? []) if (task.usage) usage = task.usage;
    if (!usage || usage.total_tokens == null) return undefined;
    return { tokens: usage.total_tokens };
  } catch {
    return undefined;
  }
}

async function readStdin(): Promise<string> {
  const chunks: Buffer[] = [];
  for await (const c of process.stdin) chunks.push(c as Buffer);
  return Buffer.concat(chunks).toString("utf-8");
}

async function main(): Promise<number> {
  const job: Job = JSON.parse((await readStdin()).split("\n")[0]);
  const scope = job.scope;

  const cp = new BedrockAgentCoreClient({ region: REGION });
  let sessionId: string | undefined;
  let browser: Browser | undefined;
  let cleanedUp = false;
  let cleanupFailed = false; // StopBrowserSession 失败 → worker 非 0 退出，让泄漏可观测（对照 Nova）

  // 会话清理（ADR 0024 终止契约，对照 Nova 的 with __exit__）：
  //   顺序——**先发 StopBrowserSession 释放会话（最重要、优先）**，再关 browser；
  //   不让易挂起的 browser.close 挟持会话释放（审计窗口 5）。close 套超时预算，避免耗尽 grace。
  //   幂等：cleanedUp 守卫，防 SIGTERM handler 与 finally 双调。
  //   StopBrowserSession 失败不静默吞：记日志 + 置 cleanupFailed → 非 0 退出（审计窗口 C）。
  async function cleanup(): Promise<void> {
    if (cleanedUp) return;
    cleanedUp = true;
    const sid = sessionId;
    if (sid) {
      try {
        await cp.send(new StopBrowserSessionCommand({ browserIdentifier: BROWSER_ID, sessionId: sid }));
      } catch (e) {
        cleanupFailed = true;
        log(`worker: StopBrowserSession FAILED (会话可能泄漏，需排查): ${(e as Error).message}`);
      }
    }
    // 会话已释放，再尽力关本地 browser；套超时，挂住也不拖垮（会话已停，close 失败无计费影响）
    if (browser) {
      await Promise.race([
        browser.close().catch(() => {}),
        new Promise<void>((r) => setTimeout(r, 3000)),
      ]);
    }
  }

  // SIGTERM：清理会话再退（防 AgentCore 会话泄漏继续烧钱）。
  let terminated = false;
  process.on("SIGTERM", async () => {
    if (terminated) return;
    terminated = true;
    log("worker: SIGTERM received, releasing AgentCore session");
    // 窗口(1) 兜底：若 SIGTERM 在 StartBrowserSession 返回前到达，sessionId 还没赋值，
    // 但服务端可能已建会话。给一小段时间让 Start 的 await 返回、sessionId 落地，再 cleanup。
    if (!sessionId) {
      await new Promise<void>((r) => setTimeout(r, 1500));
    }
    await cleanup();
    log(`worker: session shutdown complete after SIGTERM${cleanupFailed ? " (WITH FAILURE)" : ""}`);
    process.exit(cleanupFailed ? 1 : 0);
  });

  const reportRefs: Array<{ granularity: string; path: string }> = [];

  try {
    const started = await cp.send(new StartBrowserSessionCommand({ browserIdentifier: BROWSER_ID, name: "worker" }));
    sessionId = started.sessionId!;
    const wsUrl = started.streams?.automationStream?.streamEndpoint!;
    const headers = await signCdpUpgrade(wsUrl);
    browser = await chromium.connectOverCDP(wsUrl, { headers });
    const ctx = browser.contexts()[0] ?? (await browser.newContext());
    const page = ctx.pages()[0] ?? (await ctx.newPage());
    const agent = new PlaywrightAgent(page, {
      generateReport: true,
      modelConfig: MODEL_CONFIG,
      createOpenAIClient: async () => new OpenAI({ baseURL: BASE_URL, apiKey: "unused", fetch: sigv4Fetch }) as any,
    });

    emit({ type: "scope_started", scopeId: scope.id });  // 三级时长起点
    // scope 内串行跑 scenarios，共享同一会话（ADR 0019/0024）
    for (const sc of job.scenarios) {
      emit({ type: "scenario_started", scenarioId: sc.id });
      const statuses: string[] = [];
      for (const step of sc.steps) {
        statuses.push(await runStep(agent, page, sc.id, step));
      }
      emit({ type: "scenario_done", scenarioId: sc.id, status: aggregate(statuses) });
    }

    // 归集原生报告（ADR 0024 reportRefs）：destroy 后 reportFile finalize，Midscene 出 1 个 html/worker（scope 级）
    await agent.destroy().catch(() => {});
    if (agent.reportFile) reportRefs.push({ granularity: "scope", path: agent.reportFile });
  } finally {
    await cleanup();
  }

  emit({ type: "scope_done", scopeId: scope.id, sessionId: sessionId ?? null, reportRefs });
  // 正常路径若会话释放失败 → 非 0 退出，让 schedule 记 error、泄漏可观测（审计窗口 C，对照 Nova）
  return cleanupFailed ? 1 : 0;
}

async function runStep(
  agent: PlaywrightAgent, page: import("playwright").Page, scenarioId: string, step: Step,
): Promise<string> {
  const { index, keyword, text } = step;
  emit({ type: "step_started", scenarioId, stepIndex: index });  // step 时长起点
  try {
    const urlMatch = URL_IN_QUOTES.exec(text);
    if (urlMatch) {
      // 内建确定性导航（ADR 0020）：抽 URL 直接 goto，不浪费 AI
      await page.goto(urlMatch[1], { waitUntil: "domcontentloaded", timeout: 60_000 });
      emit({ type: "step_done", scenarioId, stepIndex: index, status: "passed" });
      return "passed";
    }
    if (keyword === "Then") {
      // AI 断言 + N 次投票（ADR 0014/0024）
      let yes = 0;
      for (let i = 0; i < VOTES; i++) if (await agent.aiBoolean(unquote(text))) yes++;
      const passed = yes > VOTES / 2;
      const ev: Record<string, unknown> = {
        type: "step_done", scenarioId, stepIndex: index,
        status: passed ? "passed" : "failed",
        votes: { yes, total: VOTES },
      };
      const cost = lastCost(agent);
      if (cost) ev.cost = cost;
      if (!passed) { ev.errorType = "assertion_failed"; ev.message = `AI 断言未过多数票（${yes}/${VOTES}）：${text}`; }
      emit(ev);
      return passed ? "passed" : "failed";
    }
    // When / Given（非 URL）→ AI 动作（无 votes）
    await agent.aiAct(unquote(text));
    const ev: Record<string, unknown> = { type: "step_done", scenarioId, stepIndex: index, status: "passed" };
    const cost = lastCost(agent);
    if (cost) ev.cost = cost;
    emit(ev);
    return "passed";
  } catch (e) {
    emit({
      type: "step_done", scenarioId, stepIndex: index,
      status: "error", errorType: "engine_error", message: `${(e as Error).name}: ${(e as Error).message}`,
    });
    return "error";
  }
}

function aggregate(statuses: string[]): string {
  if (statuses.includes("error")) return "error";
  if (statuses.includes("failed")) return "failed";
  return "passed";
}

main().then((code) => process.exit(code)).catch((e) => { log(`worker fatal: ${e}`); process.exit(1); });
